source/extensions/transport_sockets/alts/config.cc

```diff
@@ -0,0 +1,136 @@
+        peers.find(value) != peers.end()) {
+      return true;
+    }
+  }
+
+  err =
+      "Couldn't find peer's service account in peer_service_accounts: " + absl::StrJoin(peers, ",");
+  return false;
+}
+
+ProtobufTypes::MessagePtr AltsTransportSocketConfigFactory::createEmptyConfigProto() {
+
+  return std::make_unique<envoy::config::transport_socket::alts::v2alpha::Alts>();
+}
+
+Network::TransportSocketFactoryPtr
+UpstreamAltsTransportSocketConfigFactory::createTransportSocketFactory(
+    const Protobuf::Message& message, Server::Configuration::TransportSocketFactoryContext&) {
+  auto config =
+      MessageUtil::downcastAndValidate<const envoy::config::transport_socket::alts::v2alpha::Alts&>(
+          message);
+
+  std::string handshaker_service = config.handshaker_service();
+  const auto& peer_service_accounts = config.peer_service_accounts();
+  std::unordered_set<std::string> peers(peer_service_accounts.cbegin(),
+                                        peer_service_accounts.cend());
+
+  HandshakeValidator validator;
+  // Skip validation if peers is empty.
+  if (!peers.empty()) {
+    validator = [peers](const tsi_peer& peer, std::string& err) {
+      return doValidate(peer, peers, err);
+    };
+  }
+
+  return std::make_unique<TsiSocketFactory>(
+      [handshaker_service](Event::Dispatcher& dispatcher) {
+        grpc_alts_credentials_options* options = grpc_alts_credentials_client_options_create();
+
+        tsi_handshaker* handshaker = nullptr;
+
+        // Specifying target name as empty since TSI won't take care of
+        // validating peer identity in this use case. The validation will be
+        // implemented in TsiSocket later.
+        alts_tsi_handshaker_create(options, "", handshaker_service.c_str(), true /* is_client */,
+                                   &handshaker);
+
+        ASSERT(handshaker != nullptr);
+
+        grpc_alts_credentials_options_destroy(options);
+
+        return std::make_unique<TsiHandshaker>(handshaker, dispatcher);
+      },
+      validator);
+}
+
+Network::TransportSocketFactoryPtr
+DownstreamAltsTransportSocketConfigFactory::createTransportSocketFactory(
+    const Protobuf::Message& message, Server::Configuration::TransportSocketFactoryContext&,
+    const std::vector<std::string>&) {
+  auto config =
+      MessageUtil::downcastAndValidate<const envoy::config::transport_socket::alts::v2alpha::Alts&>(
+          message);
+
+  std::string handshaker_service = config.handshaker_service();
+  const auto& peer_service_accounts = config.peer_service_accounts();
+  std::unordered_set<std::string> peers(peer_service_accounts.cbegin(),
+                                        peer_service_accounts.cend());
+
+  HandshakeValidator validator;
+  // Skip validation if peers is empty.
+  if (!peers.empty()) {
+    validator = [peers](const tsi_peer& peer, std::string& err) {
+      return doValidate(peer, peers, err);
+    };
```

Comment:
would it make sense to have a helper function to where upstream and downstream do the asme thing?